File: app/src/main/java/uk/co/ourfriendirony/medianotifier/activities/ActivityLibrary.java
package uk.co.ourfriendirony.medianotifier.activities;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import java.util.List;

import uk.co.ourfriendirony.medianotifier.R;
import uk.co.ourfriendirony.medianotifier.activities.async.ListChildren;
import uk.co.ourfriendirony.medianotifier.activities.async.UpdateMediaItem;
import uk.co.ourfriendirony.medianotifier.activities.viewadapter.ListAdapterSummary;
import uk.co.ourfriendirony.medianotifier.clients.Client;
import uk.co.ourfriendirony.medianotifier.clients.ClientFactory;
import uk.co.ourfriendirony.medianotifier.db.Database;
import uk.co.ourfriendirony.medianotifier.db.DatabaseFactory;
import uk.co.ourfriendirony.medianotifier.general.IntentGenerator;
import uk.co.ourfriendirony.medianotifier.mediaitem.MediaItem;

import static uk.co.ourfriendirony.medianotifier.general.Constants.INTENT_KEY;

public class ActivityLibrary extends AppCompatActivity {
    private Spinner spinnerView;
    private ListView listView;
    private List<MediaItem> mediaItems;
    private ProgressBar progressBar;
    private int currentItemPos;
    private Database db;
    private Client client;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_list);

        String intentKey = getIntent().getExtras().getString(INTENT_KEY);
        getSupportActionBar().setTitle(intentKey + " Library");
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        db = new DatabaseFactory().getDatabase(getBaseContext(), intentKey);
        client = new ClientFactory().getClient(intentKey);

        spinnerView = findViewById(R.id.spinner);
        listView = findViewById(R.id.list);
        progressBar = findViewById(R.id.progress);
        spinnerView.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int itemPos, long id) {
                currentItemPos = itemPos;
                new ListChildren(parent.getContext(), progressBar, listView, db)
                        .executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, mediaItems.get(itemPos).getId());
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });

        mediaItems = db.readAllParentItems();
        if (mediaItems.size() > 0) {
            progressBar.setIndeterminate(true);
            ArrayAdapter<?> listAdapterSummary = new ListAdapterSummary(getBaseContext(), R.layout.list_item_generic_title, mediaItems, db);
            spinnerView.setAdapter(listAdapterSummary);
        } else {
            findViewById(R.id.spinner_progress).setLayoutParams(new RelativeLayout.LayoutParams(0, 0));
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_library, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem menuItem) {
        if (mediaItems.size() > 0) {
            MediaItem mediaItem = mediaItems.get(currentItemPos);
            if (menuItem.getItemId() == R.id.action_refresh) {
                new UpdateMediaItem(ActivityLibrary.this, progressBar, db, client).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, mediaItem);
                new ListChildren(ActivityLibrary.this, progressBar, listView, db).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, mediaItem.getId());
                return true;
            } else if (menuItem.getItemId() == R.id.action_remove) {
                db.delete(mediaItem.getId());
                this.recreate();
                return true;
            } else if (menuItem.getItemId() == R.id.action_lookup) {
                if (mediaItem.getExternalLink() != null) {
                    Intent intent = IntentGenerator.getWebPageIntent(mediaItem.getExternalLink());
                    startActivity(intent);
                } else {
                    Toast.makeText(this, "No External Link", Toast.LENGTH_SHORT).show();
                }
                return true;
            } else {
                return super.onOptionsItemSelected(menuItem);
            }
        }
        return false;
    }
}

File: app/src/main/kotlin/uk/co/ourfriendirony/medianotifier/activities/ActivityLibrary.kt
package uk.co.ourfriendirony.medianotifier.activities

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import uk.co.ourfriendirony.medianotifier.R
import uk.co.ourfriendirony.medianotifier.activities.async.ListChildren
import uk.co.ourfriendirony.medianotifier.activities.async.UpdateMediaItem
import uk.co.ourfriendirony.medianotifier.activities.viewadapter.ListAdapterSummary
import uk.co.ourfriendirony.medianotifier.clients.Client
import uk.co.ourfriendirony.medianotifier.clients.ClientFactory
import uk.co.ourfriendirony.medianotifier.db.Database
import uk.co.ourfriendirony.medianotifier.db.DatabaseFactory
import uk.co.ourfriendirony.medianotifier.general.Constants.INTENT_KEY
import uk.co.ourfriendirony.medianotifier.general.IntentGenerator.getWebPageIntent
import uk.co.ourfriendirony.medianotifier.mediaitem.MediaItem
import java.util.concurrent.Executors

class ActivityLibrary : AppCompatActivity() {
    private var spinnerView: Spinner? = null
    private var listView: ListView? = null
    private var mediaItems: List<MediaItem?>? = null
    private var progressBar: ProgressBar? = null
    private var currentItemPos = 0
    private var db: Database? = null
    private var client: Client? = null

    private val myHandler = Handler(Looper.getMainLooper())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_list)
        val intentKey = intent.extras!!.getString(INTENT_KEY)
        supportActionBar!!.title = "$intentKey Library"
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        db = DatabaseFactory().getDatabase(baseContext, intentKey!!)
        client = ClientFactory().getClient(intentKey)
        spinnerView = findViewById(R.id.spinner)
        listView = findViewById(R.id.list)
        progressBar = findViewById(R.id.progress)
        spinnerView!!.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(parent: AdapterView<*>?) {}
            override fun onItemSelected(parent: AdapterView<*>, view: View, itemPos: Int, id: Long) {
                currentItemPos = itemPos
                Executors.newSingleThreadExecutor().execute(
                    ListChildren(parent.context, progressBar, listView, db, myHandler, mediaItems!![itemPos]!!.id)
                )
            }
        }
        mediaItems = db!!.readAllParentItems()
        if (mediaItems!!.isNotEmpty()) {
            progressBar!!.isIndeterminate = true
            val listAdapterSummary: ArrayAdapter<*> = ListAdapterSummary(baseContext, R.layout.list_item_generic_title, mediaItems!!, db)
            spinnerView!!.adapter = listAdapterSummary
        } else {
            findViewById<View>(R.id.spinner_progress).layoutParams = RelativeLayout.LayoutParams(0, 0)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_library, menu)
        return true
    }

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        if (mediaItems!!.isNotEmpty()) {
            val mediaItem = mediaItems!![currentItemPos]
            return when (menuItem.itemId) {
                R.id.action_refresh -> {
                    Executors.newSingleThreadExecutor().execute(
                        UpdateMediaItem(this@ActivityLibrary, progressBar, db, client, myHandler, mediaItem!!)
                    )
                    Executors.newSingleThreadExecutor().execute(
                        ListChildren(this@ActivityLibrary, progressBar, listView, db, myHandler, mediaItem.id)
                    )
                    true
                }
                R.id.action_remove -> {
                    db!!.delete(mediaItem!!.id)
                    recreate()
                    true
                }
                R.id.action_lookup -> {
                    if (mediaItem!!.externalLink != null) {
                        val intent = getWebPageIntent(mediaItem.externalLink)
                        startActivity(intent)
                    } else {
                        Toast.makeText(this, "No External Link", Toast.LENGTH_SHORT).show()
                    }
                    true
                }
                else -> {
                    super.onOptionsItemSelected(menuItem)
                }
            }
        }
        return false
    }
}